<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html><head><meta http-equiv="Content-Type" content="text/html;charset=UTF-8">
<title>RNAstructure Classes: Dynalign_object Class Reference</title>
<link href="doxygen.css" rel="stylesheet" type="text/css">
<link href="tabs.css" rel="stylesheet" type="text/css">
</head><body>
<!-- Generated by Doxygen 1.5.7.1 -->
<div class="navigation" id="top">
  <div class="tabs">
    <ul>
      <li><a href="index.html"><span>Main&nbsp;Page</span></a></li>
      <li class="current"><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="files.html"><span>Files</span></a></li>
    </ul>
  </div>
  <div class="tabs">
    <ul>
      <li><a href="annotated.html"><span>Class&nbsp;List</span></a></li>
      <li><a href="hierarchy.html"><span>Class&nbsp;Hierarchy</span></a></li>
      <li><a href="functions.html"><span>Class&nbsp;Members</span></a></li>
    </ul>
  </div>
</div>
<div class="contents">
<h1>Dynalign_object Class Reference</h1><!-- doxytag: class="Dynalign_object" --><!-- doxytag: inherits="TwoRNA" --><a class="el" href="class_dynalign__object.html" title="Dynalign_object Class.">Dynalign_object</a> Class.  
<a href="#_details">More...</a>
<p>
<code>#include &lt;Dynalign_object.h&gt;</code>
<p>
Inherits <a class="el" href="class_two_r_n_a.html">TwoRNA</a>.
<p>

<p>
<a href="class_dynalign__object-members.html">List of all members.</a><table border="0" cellpadding="0" cellspacing="0">
<tr><td></td></tr>
<tr><td colspan="2"><br><h2>Public Member Functions</h2></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_dynalign__object.html#d578dcde5db14809aa988655e48aaf8a">Dynalign_object</a> ()</td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_dynalign__object.html#47a38b5e8cd54b68148b0b10385472a1">Dynalign_object</a> (const char sequence1[], const char sequence2[], const bool IsRNA=true)</td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_dynalign__object.html#7d60269bb4d78997c8f36793f9458f16">Dynalign_object</a> (const char filename1[], const int type1, const char filename2[], const int type2, const bool IsRNA=true)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Constructor.  <a href="#7d60269bb4d78997c8f36793f9458f16"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_dynalign__object.html#04d3f9d3e8f1b49167f7f431c0d533c6">Dynalign_object</a> (const char filename[])</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Constructor.  <a href="#04d3f9d3e8f1b49167f7f431c0d533c6"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_dynalign__object.html#859e83f3f59cb868783f0b05fb7afd25">Dynalign_object</a> (const char *filename, const short maxtrace, const short bpwin, const short awin, const short percent)</td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">int&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_dynalign__object.html#7d66a8c03191eefc52599a6f78a7f53c">Dynalign</a> (const short int maxtrace, const short int bpwin, const short int awin, const short int percent, const short int imaxseparation=-99, const float <a class="el" href="class_dynalign__object.html#4d6d04c0f6e6574db9eb38626aa856e4">gap</a>=0.4, const bool singleinsert=true, const char savefile[]=NULL, const bool optimalonly=false, const short int singlefold_subopt_percent=30, const bool local=false, const short int numProcessors=1, const int maxpairs=-1)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Predict the lowest free energy structure common to two sequences and suboptimal solutions with the Dynalign algorithm.  <a href="#7d66a8c03191eefc52599a6f78a7f53c"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_dynalign__object.html#b45eca04479d94532af1ac819b5311c7">WriteAlignment</a> (const char filename[])</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Write the alignment to disk.  <a href="#b45eca04479d94532af1ac819b5311c7"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">int&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_dynalign__object.html#038230c0e6d77ebc75e9591c00e4f472">ForceAlignment</a> (const int i, const int k)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Force an alignment during a Dynalign calculation).  <a href="#038230c0e6d77ebc75e9591c00e4f472"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">int&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_dynalign__object.html#f8e72c63d3daa80f8c3e9f53b6390a15">GetForcedAlignment</a> (const int i, const int seq)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Get an alignment constraint.  <a href="#f8e72c63d3daa80f8c3e9f53b6390a15"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">int&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_dynalign__object.html#e3026dd8710294fff44075beb1d87ba5">ReadAlignmentConstraints</a> (const char filename[])</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Read alignment constraints from disk.  <a href="#e3026dd8710294fff44075beb1d87ba5"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">int&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_dynalign__object.html#65b0d41d9ab5744f9e85710a569307b4">Templatefromct</a> (const char ctfilename[])</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Read a ct file to determine what pairs will be allowed for sequence 1 in a subsequent dynalign calculation.  <a href="#65b0d41d9ab5744f9e85710a569307b4"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">int&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_dynalign__object.html#8527a9e588154df8a2be9de660e34f78">Templatefromdsv</a> (const char dsvfilename[], const float maxdsvchange)</td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">double&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_dynalign__object.html#d6b4b08636ae75519363a01670a5a730">GetBestPairEnergy</a> (const int sequence, const int i, const int j)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Report the best energy for pair i-j from sequence number sequence (1 or 2).  <a href="#d6b4b08636ae75519363a01670a5a730"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">double&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_dynalign__object.html#0b21e9b2eb0df4f07abd9b919bac6de1">GetLowestEnergy</a> ()</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Report the lowest total free energy change from a Dynalign calculation.  <a href="#0b21e9b2eb0df4f07abd9b919bac6de1"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">char *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_dynalign__object.html#2afa4d142468d28e8a21a69214788e1b">GetErrorMessage</a> (const int error)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Return error messages based on code from GetErrorCode and other error codes.  <a href="#2afa4d142468d28e8a21a69214788e1b"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_dynalign__object.html#91cb3d5d954bd38a064cd125588c90a6">SetProgress</a> (TProgressDialog &amp;Progress)</td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_dynalign__object.html#341992ef072d745376e5011829b93b42">StopProgress</a> ()</td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_dynalign__object.html#7a9094be4ca82ff0d84aa8cd5516cdc4">~Dynalign_object</a> ()</td></tr>

<tr><td colspan="2"><br><h2>Private Member Functions</h2></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_dynalign__object.html#08df08aef1ea8e719f3fbf3350b6b37e">CommonConstructor</a> ()</td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_dynalign__object.html#9cb25652ef83c0ea95ede44bf23e4334">AllocateForceAlign</a> ()</td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_dynalign__object.html#6ae440fbbbeb2d5affd4eecc7d35a854">storetemplatefilename</a> (const char *name)</td></tr>

<tr><td colspan="2"><br><h2>Private Attributes</h2></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">short **&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_dynalign__object.html#08dd429e7365f45f0af4ac65f91a38b3">align</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">short **&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_dynalign__object.html#a71db4fdbda2e78a7aad59461ab1e499">forcealign</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">bool&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_dynalign__object.html#51fda66b732a81b71221c1025535a389">dsv_templated</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">bool&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_dynalign__object.html#57fb75d0ae9ae4d7ddb660ef46073304">ct_templated</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">char *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_dynalign__object.html#ac369239b115d06378f886687438cf46">templatefilename</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">float&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_dynalign__object.html#fdf4fc676f616371f56e4c9fef9df01f">MAXDSV</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">int&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_dynalign__object.html#d0016552f28504bf8fb4e514d493e68b">modificationflag</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">dynalignarray *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_dynalign__object.html#f4e94d7ad0c260badd19bc00c9c9c9b6">w</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">dynalignarray *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_dynalign__object.html#6fcebe057ab0c01d452f0f9101494da5">vmod</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">varray *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_dynalign__object.html#8ca9c12e9cb5f53dac734e150d5b39be">v</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">wendarray *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_dynalign__object.html#9703f79b7dd3d5349ffcad85a02a87b5">w5</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">wendarray *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_dynalign__object.html#a2809c6e4cb309f6598a9f3cc93268f8">w3</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">short *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_dynalign__object.html#aeb0bf4251ffad85d88791409b74438e">lowend</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">short *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_dynalign__object.html#85bf2b95d7c490f01c0b3c63d1391696">highend</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">datatable *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_dynalign__object.html#a1e04d4bab3329f52d69487d6deb87d0">data</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">short&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_dynalign__object.html#4d6d04c0f6e6574db9eb38626aa856e4">gap</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">short&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_dynalign__object.html#46b200ee4d625bf5b93dd25ee80282e5">lowest</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">int&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_dynalign__object.html#26f77a6eaa11a89728b53e96c09d25e5">Maxtrace</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">bool&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_dynalign__object.html#092a0c489e8c0cd1453e6c593ac25240">savefileread</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">double ***&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_dynalign__object.html#a0ef3ac79a7e95a7f68357cd7dd7d7a2">array</a></td></tr>

</table>
<hr><a name="_details"></a><h2>Detailed Description</h2>
<a class="el" href="class_dynalign__object.html" title="Dynalign_object Class.">Dynalign_object</a> Class. 
<p>
The <a class="el" href="class_dynalign__object.html" title="Dynalign_object Class.">Dynalign_object</a> class provides an entry point for the Dynalign algorithm. The class is inherited from the <a class="el" href="class_two_r_n_a.html" title="TwoRNA Class.">TwoRNA</a> class, which itself contains two instances to the class <a class="el" href="class_r_n_a.html" title="RNA Class.">RNA</a>. <hr><h2>Constructor &amp; Destructor Documentation</h2>
<a class="anchor" name="d578dcde5db14809aa988655e48aaf8a"></a><!-- doxytag: member="Dynalign_object::Dynalign_object" ref="d578dcde5db14809aa988655e48aaf8a" args="()" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">Dynalign_object::Dynalign_object           </td>
          <td>(</td>
          <td class="paramname">          </td>
          <td>&nbsp;)&nbsp;</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Constructor This is a default constructor that calls the <a class="el" href="class_two_r_n_a.html" title="TwoRNA Class.">TwoRNA</a> default constructor. 
</div>
</div><p>
<a class="anchor" name="47a38b5e8cd54b68148b0b10385472a1"></a><!-- doxytag: member="Dynalign_object::Dynalign_object" ref="47a38b5e8cd54b68148b0b10385472a1" args="(const char sequence1[], const char sequence2[], const bool IsRNA=true)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">Dynalign_object::Dynalign_object           </td>
          <td>(</td>
          <td class="paramtype">const char&nbsp;</td>
          <td class="paramname"> <em>sequence1</em>[], </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char&nbsp;</td>
          <td class="paramname"> <em>sequence2</em>[], </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const bool&nbsp;</td>
          <td class="paramname"> <em>IsRNA</em> = <code>true</code></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
This constuctor is available to reach the <a class="el" href="class_two_r_n_a.html" title="TwoRNA Class.">TwoRNA</a> constructors, from which this class is inherited. This constructor uses two cstrings to provide sequences. IsRNA is true for <a class="el" href="class_r_n_a.html" title="RNA Class.">RNA</a> folding and flase for DNA. Constructor This constuctor is available to reach the <a class="el" href="class_two_r_n_a.html" title="TwoRNA Class.">TwoRNA</a> constructors, from which this class is inherited. This constructor uses two cstrings to provide sequences. IsRNA is true for <a class="el" href="class_r_n_a.html" title="RNA Class.">RNA</a> folding and false for DNA. Input sequences should contain A,C,G,T,U,a,c,g,t,u,x,X. Capitalization makes no difference. T=t=u=U. If IsRNA is true, the backbone is <a class="el" href="class_r_n_a.html" title="RNA Class.">RNA</a>, so U is assumed. If IsRNA is false, the backbone is DNA, so T is assumed. x=X= nucleotide that neither stacks nor pairs. For now, any unknown nuc is considered 'X'. Both sequences are passed to underlying <a class="el" href="class_r_n_a.html" title="RNA Class.">RNA</a> classes for each sequence. <dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>sequence1</em>&nbsp;</td><td>is a NULL terminated c string for sequence 1. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>sequence2</em>&nbsp;</td><td>is a NULL terminated c string for sequence 2. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>IsRNA</em>&nbsp;</td><td>is a bool that indicates whether these sequences are <a class="el" href="class_r_n_a.html" title="RNA Class.">RNA</a> or DNA. true=RNA. false=DNA. Default is true. Both sequences must have the same backbone. </td></tr>
  </table>
</dl>

</div>
</div><p>
<a class="anchor" name="7d60269bb4d78997c8f36793f9458f16"></a><!-- doxytag: member="Dynalign_object::Dynalign_object" ref="7d60269bb4d78997c8f36793f9458f16" args="(const char filename1[], const int type1, const char filename2[], const int type2, const bool IsRNA=true)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">Dynalign_object::Dynalign_object           </td>
          <td>(</td>
          <td class="paramtype">const char&nbsp;</td>
          <td class="paramname"> <em>filename1</em>[], </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int&nbsp;</td>
          <td class="paramname"> <em>type1</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char&nbsp;</td>
          <td class="paramname"> <em>filename2</em>[], </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int&nbsp;</td>
          <td class="paramname"> <em>type2</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const bool&nbsp;</td>
          <td class="paramname"> <em>IsRNA</em> = <code>true</code></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Constructor. 
<p>
This constuctor is available to reach the <a class="el" href="class_two_r_n_a.html" title="TwoRNA Class.">TwoRNA</a> constructors, from which this class is inherited. This constructor uses two ctsirngs as filenames, accompanied by integers to set the file type. IsRNA is true for <a class="el" href="class_r_n_a.html" title="RNA Class.">RNA</a> folding and false for DNA. The existing files, specified by filenames, can either be a ct file, a sequence, or an RNAstructure save file. Therefore, the user provides a flag for the file type: type = 1 =&gt; .ct file, type = 2 =&gt; .seq file, type = 3 =&gt; partition function save (.pfs) file, type = 4 =&gt; folding save file (.sav). The file opening is performed by the constructors for the <a class="el" href="class_r_n_a.html" title="RNA Class.">RNA</a> classes that underlie each sequence. This constructor generates internal error codes that can be accessed by <a class="el" href="class_two_r_n_a.html#ffe4134ad81410046302219e191ba9f9" title="Return an error code, where a return of zero is no error.">GetErrorCode()</a> after the constructor is called. 0 = no error. The errorcode can be resolved to a c string using GetErrorMessage. Note that the contructor needs to be explicitly told, via IsRNA, what the backbone is because files do not store this information. Note also that save files explicitly store the thermodynamic parameters, therefore changing the backbone type as compaared to the original calculation will not change structure predictions. <dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>filename1</em>&nbsp;</td><td>is a null terminated c string and refers to sequence 1. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>filename2</em>&nbsp;</td><td>is a null terminated c string and refers to sequence 2. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>type1</em>&nbsp;</td><td>is an integer that indicates the file type for sequence 1. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>type2</em>&nbsp;</td><td>is an integer that indicates the file type for sequence 2. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>IsRNA</em>&nbsp;</td><td>is a bool that indicates whether these sequences are <a class="el" href="class_r_n_a.html" title="RNA Class.">RNA</a> or DNA. true=RNA. false=DNA. Default is true. Only one backbone is allowed for both sequences. </td></tr>
  </table>
</dl>

</div>
</div><p>
<a class="anchor" name="04d3f9d3e8f1b49167f7f431c0d533c6"></a><!-- doxytag: member="Dynalign_object::Dynalign_object" ref="04d3f9d3e8f1b49167f7f431c0d533c6" args="(const char filename[])" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">Dynalign_object::Dynalign_object           </td>
          <td>(</td>
          <td class="paramtype">const char&nbsp;</td>
          <td class="paramname"> <em>filename</em>[]          </td>
          <td>&nbsp;)&nbsp;</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Constructor. 
<p>
This constructor allows the user to read a dynalign save file (.dsv) to get base pairing information. This constructor generates internal error codes that can be accessed by <a class="el" href="class_two_r_n_a.html#ffe4134ad81410046302219e191ba9f9" title="Return an error code, where a return of zero is no error.">GetErrorCode()</a> after the constructor is called. 0 = no error. The errorcode can be resolved to a c string using GetErrorMessage. <dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>filename[]</em>&nbsp;</td><td>is a cstring that indicates the filename of a .dsv file. </td></tr>
  </table>
</dl>

</div>
</div><p>
<a class="anchor" name="859e83f3f59cb868783f0b05fb7afd25"></a><!-- doxytag: member="Dynalign_object::Dynalign_object" ref="859e83f3f59cb868783f0b05fb7afd25" args="(const char *filename, const short maxtrace, const short bpwin, const short awin, const short percent)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">Dynalign_object::Dynalign_object           </td>
          <td>(</td>
          <td class="paramtype">const char *&nbsp;</td>
          <td class="paramname"> <em>filename</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const short&nbsp;</td>
          <td class="paramname"> <em>maxtrace</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const short&nbsp;</td>
          <td class="paramname"> <em>bpwin</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const short&nbsp;</td>
          <td class="paramname"> <em>awin</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const short&nbsp;</td>
          <td class="paramname"> <em>percent</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Constructor This constructor is used to perform Dynaligh refolding. This does not allow any changes in constraints, but does allow the creation of different set of suboptimal structures. This constructor generates internal error codes that can be accessed by <a class="el" href="class_two_r_n_a.html#ffe4134ad81410046302219e191ba9f9" title="Return an error code, where a return of zero is no error.">GetErrorCode()</a> after the constructor is called. 0 = no error. The errorcode can be resolved to a c string using GetErrorMessage. <dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>filename</em>&nbsp;</td><td>is the name of a Dynalign save file name (.dsv). </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>maxtrace</em>&nbsp;</td><td>is the maximum number of common structures to be determined. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>bpwin</em>&nbsp;</td><td>the the base pair window parameter, where 0 allows the structures to have similar pairs and larger windows make the structures more diverse. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>awin</em>&nbsp;</td><td>is the alignment window parameter, where 0 allows the alignments to be similar and larger values make the alignments more diverse. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>percent</em>&nbsp;</td><td>is the maximum percent difference in total folding free energy change above the lowest for suboptimal common structures. </td></tr>
  </table>
</dl>

</div>
</div><p>
<a class="anchor" name="7a9094be4ca82ff0d84aa8cd5516cdc4"></a><!-- doxytag: member="Dynalign_object::~Dynalign_object" ref="7a9094be4ca82ff0d84aa8cd5516cdc4" args="()" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">Dynalign_object::~Dynalign_object           </td>
          <td>(</td>
          <td class="paramname">          </td>
          <td>&nbsp;)&nbsp;</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>

</div>
</div><p>
<hr><h2>Member Function Documentation</h2>
<a class="anchor" name="9cb25652ef83c0ea95ede44bf23e4334"></a><!-- doxytag: member="Dynalign_object::AllocateForceAlign" ref="9cb25652ef83c0ea95ede44bf23e4334" args="()" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Dynalign_object::AllocateForceAlign           </td>
          <td>(</td>
          <td class="paramname">          </td>
          <td>&nbsp;)&nbsp;</td>
          <td><code> [private]</code></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>

</div>
</div><p>
<a class="anchor" name="08df08aef1ea8e719f3fbf3350b6b37e"></a><!-- doxytag: member="Dynalign_object::CommonConstructor" ref="08df08aef1ea8e719f3fbf3350b6b37e" args="()" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Dynalign_object::CommonConstructor           </td>
          <td>(</td>
          <td class="paramname">          </td>
          <td>&nbsp;)&nbsp;</td>
          <td><code> [private]</code></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>

</div>
</div><p>
<a class="anchor" name="7d66a8c03191eefc52599a6f78a7f53c"></a><!-- doxytag: member="Dynalign_object::Dynalign" ref="7d66a8c03191eefc52599a6f78a7f53c" args="(const short int maxtrace, const short int bpwin, const short int awin, const short int percent, const short int imaxseparation=&#45;99, const float gap=0.4, const bool singleinsert=true, const char savefile[]=NULL, const bool optimalonly=false, const short int singlefold_subopt_percent=30, const bool local=false, const short int numProcessors=1, const int maxpairs=&#45;1)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int Dynalign_object::Dynalign           </td>
          <td>(</td>
          <td class="paramtype">const short int&nbsp;</td>
          <td class="paramname"> <em>maxtrace</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const short int&nbsp;</td>
          <td class="paramname"> <em>bpwin</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const short int&nbsp;</td>
          <td class="paramname"> <em>awin</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const short int&nbsp;</td>
          <td class="paramname"> <em>percent</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const short int&nbsp;</td>
          <td class="paramname"> <em>imaxseparation</em> = <code>-99</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const float&nbsp;</td>
          <td class="paramname"> <em>gap</em> = <code>0.4</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const bool&nbsp;</td>
          <td class="paramname"> <em>singleinsert</em> = <code>true</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char&nbsp;</td>
          <td class="paramname"> <em>savefile</em>[] = <code>NULL</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const bool&nbsp;</td>
          <td class="paramname"> <em>optimalonly</em> = <code>false</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const short int&nbsp;</td>
          <td class="paramname"> <em>singlefold_subopt_percent</em> = <code>30</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const bool&nbsp;</td>
          <td class="paramname"> <em>local</em> = <code>false</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const short int&nbsp;</td>
          <td class="paramname"> <em>numProcessors</em> = <code>1</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int&nbsp;</td>
          <td class="paramname"> <em>maxpairs</em> = <code>-1</code></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Predict the lowest free energy structure common to two sequences and suboptimal solutions with the Dynalign algorithm. 
<p>
In case of error, the function returns a non-zero that can be parsed by <a class="el" href="class_dynalign__object.html#2afa4d142468d28e8a21a69214788e1b" title="Return error messages based on code from GetErrorCode and other error codes.">GetErrorMessage()</a> or <a class="el" href="class_two_r_n_a.html#11f60c73bc4b874123376a989cd7f6a2" title="Return error messages based on code from GetErrorCode and other error codes.">GetErrorMessageString()</a>. <dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>maxtrace</em>&nbsp;</td><td>is the maximum number of common structures to be determined. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>bpwin</em>&nbsp;</td><td>the the base pair window parameter, where 0 allows the structures to have similar pairs and larger windows make the structures more diverse. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>awin</em>&nbsp;</td><td>is the alignment window parameter, where 0 allows the alignments to be similar and larger values make the alignments more diverse. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>percent</em>&nbsp;</td><td>is the maximum percent difference in total folding free energy change above the lowest for suboptimal common structures. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>imaxseparation</em>&nbsp;</td><td>is the maximum separation between aligned nucleotides. Values &gt;= 0 are the traditional parameter, those below zero trigger the HMM alignment method, which is now prefered. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>gap</em>&nbsp;</td><td>is the cost of adding gap nucleotides in the alignment in kcal/mol. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>singleinsert</em>&nbsp;</td><td>is whether single basepair inserts are allowed in one sequence vs the other. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>savefile</em>&nbsp;</td><td>is c-string with the name of a dynalign savefile (*.dsv) to be created. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>optimalonly</em>&nbsp;</td><td>can be used to turn on a calculation of only the energy (when true) and not the structures. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>singlefold_subopt_percent</em>&nbsp;</td><td>is the maximum % difference of folding energy above the lowest free energy structure for pairs in single sequence folding that will be allowed in the dynalign calculation. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>local</em>&nbsp;</td><td>is whether Dynalign is being run in local (true) or global mode (false). </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>numProcessors</em>&nbsp;</td><td>is the number of processors to use for the calculation. This requires a compilation for SMP. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>maxpairs</em>&nbsp;</td><td>is under development for multiple sequence folding. Use -1 (default) for now. </td></tr>
  </table>
</dl>
<dl class="return" compact><dt><b>Returns:</b></dt><dd>An int that indicates an error code (0 = no error, non-zero = error occurred). </dd></dl>

</div>
</div><p>
<a class="anchor" name="038230c0e6d77ebc75e9591c00e4f472"></a><!-- doxytag: member="Dynalign_object::ForceAlignment" ref="038230c0e6d77ebc75e9591c00e4f472" args="(const int i, const int k)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int Dynalign_object::ForceAlignment           </td>
          <td>(</td>
          <td class="paramtype">const int&nbsp;</td>
          <td class="paramname"> <em>i</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int&nbsp;</td>
          <td class="paramname"> <em>k</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Force an alignment during a Dynalign calculation). 
<p>
Nucleotide i from sequence 1 will be aligned to nucleotide k in sequence 2 in subsequent Dynalign calculation. The function returns 0 with no error and a non-zero otherwise that can be parsed by <a class="el" href="class_dynalign__object.html#2afa4d142468d28e8a21a69214788e1b" title="Return error messages based on code from GetErrorCode and other error codes.">GetErrorMessage()</a> or <a class="el" href="class_two_r_n_a.html#11f60c73bc4b874123376a989cd7f6a2" title="Return error messages based on code from GetErrorCode and other error codes.">GetErrorMessageString()</a>. <dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>i</em>&nbsp;</td><td>is the index of nucleotide from sequence 1. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>k</em>&nbsp;</td><td>is the index of nucleotide from sequence 2. </td></tr>
  </table>
</dl>
<dl class="return" compact><dt><b>Returns:</b></dt><dd>An integer that indicates an error code (0 = no error, 100 = nucleotide i out of range, 101 = nucleotide k out of range). </dd></dl>

</div>
</div><p>
<a class="anchor" name="d6b4b08636ae75519363a01670a5a730"></a><!-- doxytag: member="Dynalign_object::GetBestPairEnergy" ref="d6b4b08636ae75519363a01670a5a730" args="(const int sequence, const int i, const int j)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double Dynalign_object::GetBestPairEnergy           </td>
          <td>(</td>
          <td class="paramtype">const int&nbsp;</td>
          <td class="paramname"> <em>sequence</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int&nbsp;</td>
          <td class="paramname"> <em>i</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int&nbsp;</td>
          <td class="paramname"> <em>j</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Report the best energy for pair i-j from sequence number sequence (1 or 2). 
<p>
This function reports the lowest ffolding free energy for any pairs between i-j in sequence number sequence (1 or 2). This requires a search over all possible pairs in the second sequence. NOTE: This function ONLY works after reading a Dynalign save file (.dsv) using the constructor. This is because the Dynalign energies are not normally stored after calling Dynalign. This function generates internal error codes that can be accessed by <a class="el" href="class_two_r_n_a.html#ffe4134ad81410046302219e191ba9f9" title="Return an error code, where a return of zero is no error.">GetErrorCode()</a> after the constructor is called. 0 = no error, 107 = Data not available, 108 = nucleotide out of range. The errorcode can be resolved to a c string using GetErrorMessage. <dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>sequence</em>&nbsp;</td><td>is an integer indicating the sequence # (must be 1 or 2). </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>i</em>&nbsp;</td><td>is the 5' nucleotide in a pair. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>j</em>&nbsp;</td><td>is the 3' nucleotide in a pair. </td></tr>
  </table>
</dl>
<dl class="return" compact><dt><b>Returns:</b></dt><dd>A double that gives an energy in kcal/mol. </dd></dl>

</div>
</div><p>
<a class="anchor" name="2afa4d142468d28e8a21a69214788e1b"></a><!-- doxytag: member="Dynalign_object::GetErrorMessage" ref="2afa4d142468d28e8a21a69214788e1b" args="(const int error)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">char * Dynalign_object::GetErrorMessage           </td>
          <td>(</td>
          <td class="paramtype">const int&nbsp;</td>
          <td class="paramname"> <em>error</em>          </td>
          <td>&nbsp;)&nbsp;</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Return error messages based on code from GetErrorCode and other error codes. 
<p>
0 = no error 100-999 = Error associated with Dynalign, to be handled here. &gt;=1000 = Errors for underlying sequence, get message from <a class="el" href="class_two_r_n_a.html" title="TwoRNA Class.">TwoRNA</a> base class. Current errors handled here are: 100 "Nucleotide from sequence 1 is out of range.\n"; 101 "Nucleotide from sequence 2 is out of range.\n"; 102 "Alignment constraint file not found.\n"; 103 "Error reading alignment constraint file.\n"; 104 "CT file not found.\n"; 105 "A template has already been specified; only one is allowed.\n"; 106 "DSV file not found.\n"; 107 "Data not available to calculate energy.\n" 108 "Nucleotide out of range.\n"; 109 "Value of maxpairs is too large to be achievable.\n" 110 "Error reading thermodynamic parameters." <dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>error</em>&nbsp;</td><td>is the integer error code provided by <a class="el" href="class_two_r_n_a.html#ffe4134ad81410046302219e191ba9f9" title="Return an error code, where a return of zero is no error.">GetErrorCode()</a> or by a call to a function that returns an error. </td></tr>
  </table>
</dl>
<dl class="return" compact><dt><b>Returns:</b></dt><dd>A pointer to a c string that provides an error message or from other functions that return integer error codes. </dd></dl>

<p>Reimplemented from <a class="el" href="class_two_r_n_a.html#073426082daa436c724225a1c9a55b67">TwoRNA</a>.</p>

</div>
</div><p>
<a class="anchor" name="f8e72c63d3daa80f8c3e9f53b6390a15"></a><!-- doxytag: member="Dynalign_object::GetForcedAlignment" ref="f8e72c63d3daa80f8c3e9f53b6390a15" args="(const int i, const int seq)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int Dynalign_object::GetForcedAlignment           </td>
          <td>(</td>
          <td class="paramtype">const int&nbsp;</td>
          <td class="paramname"> <em>i</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int&nbsp;</td>
          <td class="paramname"> <em>seq</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Get an alignment constraint. 
<p>
<dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>i</em>&nbsp;</td><td>is the nucleotide number. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>seq</em>&nbsp;</td><td>is the sequence (1 or 2) from which i is derived. </td></tr>
  </table>
</dl>
<dl class="return" compact><dt><b>Returns:</b></dt><dd>An integer that indicates the nucleotide to which i is forced to be aligned, where 0 indicates no alignment. </dd></dl>

</div>
</div><p>
<a class="anchor" name="0b21e9b2eb0df4f07abd9b919bac6de1"></a><!-- doxytag: member="Dynalign_object::GetLowestEnergy" ref="0b21e9b2eb0df4f07abd9b919bac6de1" args="()" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double Dynalign_object::GetLowestEnergy           </td>
          <td>(</td>
          <td class="paramname">          </td>
          <td>&nbsp;)&nbsp;</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Report the lowest total free energy change from a Dynalign calculation. 
<p>
NOTE: This function ONLY works after reading a Dynalign save file (.dsv) using the constructor. This is because the Dynalign energies are not normally stored after calling Dynalign. This function generates internal error codes that can be accessed by <a class="el" href="class_two_r_n_a.html#ffe4134ad81410046302219e191ba9f9" title="Return an error code, where a return of zero is no error.">GetErrorCode()</a> after the constructor is called. 0 = no error, 107 = Data not available, 108 = nucleotide out of range. The errorcode can be resolved to a c string using GetErrorMessage. <dl class="return" compact><dt><b>Returns:</b></dt><dd>a double that gives an energy in kcal/mol. </dd></dl>

</div>
</div><p>
<a class="anchor" name="e3026dd8710294fff44075beb1d87ba5"></a><!-- doxytag: member="Dynalign_object::ReadAlignmentConstraints" ref="e3026dd8710294fff44075beb1d87ba5" args="(const char filename[])" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int Dynalign_object::ReadAlignmentConstraints           </td>
          <td>(</td>
          <td class="paramtype">const char&nbsp;</td>
          <td class="paramname"> <em>filename</em>[]          </td>
          <td>&nbsp;)&nbsp;</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Read alignment constraints from disk. 
<p>
The file format is: i1 k1 i2 k2 -1 -1 Where each line gives a aligned pair (i from sequence 1 and k from sequence 2). The file terminates with -1 -1 to indicate the file end. The function returns 0 with no error and a non-zero otherwise that can be parsed by <a class="el" href="class_dynalign__object.html#2afa4d142468d28e8a21a69214788e1b" title="Return error messages based on code from GetErrorCode and other error codes.">GetErrorMessage()</a> or <a class="el" href="class_two_r_n_a.html#11f60c73bc4b874123376a989cd7f6a2" title="Return error messages based on code from GetErrorCode and other error codes.">GetErrorMessageString()</a>. <dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>filename</em>&nbsp;</td><td>is a c string that is the file name to be read. </td></tr>
  </table>
</dl>
<dl class="return" compact><dt><b>Returns:</b></dt><dd>An integer that indicates an error code (0 = no error, 102 = file not found, 103 = error reading constraint file). </dd></dl>

</div>
</div><p>
<a class="anchor" name="91cb3d5d954bd38a064cd125588c90a6"></a><!-- doxytag: member="Dynalign_object::SetProgress" ref="91cb3d5d954bd38a064cd125588c90a6" args="(TProgressDialog &amp;Progress)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Dynalign_object::SetProgress           </td>
          <td>(</td>
          <td class="paramtype">TProgressDialog &amp;&nbsp;</td>
          <td class="paramname"> <em>Progress</em>          </td>
          <td>&nbsp;)&nbsp;</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Provide a TProgressDialog for following calculation progress. A TProgressDialog class has a public function void update(int percent) that indicates the progress of a long calculation. <dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>Progress</em>&nbsp;</td><td>is a TProgressDialog class. </td></tr>
  </table>
</dl>

</div>
</div><p>
<a class="anchor" name="341992ef072d745376e5011829b93b42"></a><!-- doxytag: member="Dynalign_object::StopProgress" ref="341992ef072d745376e5011829b93b42" args="()" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Dynalign_object::StopProgress           </td>
          <td>(</td>
          <td class="paramname">          </td>
          <td>&nbsp;)&nbsp;</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Provide a means to stop using a TProgressDialog. StopProgress tells the <a class="el" href="class_r_n_a.html" title="RNA Class.">RNA</a> class to no longer follow progress. This should be called if the TProgressDialog is deleted, so that this class does not make reference to it. 
</div>
</div><p>
<a class="anchor" name="6ae440fbbbeb2d5affd4eecc7d35a854"></a><!-- doxytag: member="Dynalign_object::storetemplatefilename" ref="6ae440fbbbeb2d5affd4eecc7d35a854" args="(const char *name)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Dynalign_object::storetemplatefilename           </td>
          <td>(</td>
          <td class="paramtype">const char *&nbsp;</td>
          <td class="paramname"> <em>name</em>          </td>
          <td>&nbsp;)&nbsp;</td>
          <td><code> [private]</code></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>

</div>
</div><p>
<a class="anchor" name="65b0d41d9ab5744f9e85710a569307b4"></a><!-- doxytag: member="Dynalign_object::Templatefromct" ref="65b0d41d9ab5744f9e85710a569307b4" args="(const char ctfilename[])" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int Dynalign_object::Templatefromct           </td>
          <td>(</td>
          <td class="paramtype">const char&nbsp;</td>
          <td class="paramname"> <em>ctfilename</em>[]          </td>
          <td>&nbsp;)&nbsp;</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Read a ct file to determine what pairs will be allowed for sequence 1 in a subsequent dynalign calculation. 
<p>
This results in all pairs but those in the ct being disallowed. <dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>ctfilename</em>&nbsp;</td><td>is the name of the ct file to be read to provide the template. </td></tr>
  </table>
</dl>
<dl class="return" compact><dt><b>Returns:</b></dt><dd>An integer that indicates an error code (0=no error, 104=file not found, 105=template is already specified) </dd></dl>

</div>
</div><p>
<a class="anchor" name="8527a9e588154df8a2be9de660e34f78"></a><!-- doxytag: member="Dynalign_object::Templatefromdsv" ref="8527a9e588154df8a2be9de660e34f78" args="(const char dsvfilename[], const float maxdsvchange)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int Dynalign_object::Templatefromdsv           </td>
          <td>(</td>
          <td class="paramtype">const char&nbsp;</td>
          <td class="paramname"> <em>dsvfilename</em>[], </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const float&nbsp;</td>
          <td class="paramname"> <em>maxdsvchange</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
This reads a dsv file and only allows pairs with folding free energy change between the lowest and lowest + maxdsvchange in a subsequent dynalign calculation. <dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>dsvfilename</em>&nbsp;</td><td>is the name of the ct file to be read to provide the template. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>maxdsvchange</em>&nbsp;</td><td>in a float that gives a percent difference in free energy above the lowest free energy change. </td></tr>
  </table>
</dl>
<dl class="return" compact><dt><b>Returns:</b></dt><dd>An integer that indicates an error code (0=no error, 106=file not found, 105=template is already specified) </dd></dl>

</div>
</div><p>
<a class="anchor" name="b45eca04479d94532af1ac819b5311c7"></a><!-- doxytag: member="Dynalign_object::WriteAlignment" ref="b45eca04479d94532af1ac819b5311c7" args="(const char filename[])" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Dynalign_object::WriteAlignment           </td>
          <td>(</td>
          <td class="paramtype">const char&nbsp;</td>
          <td class="paramname"> <em>filename</em>[]          </td>
          <td>&nbsp;)&nbsp;</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Write the alignment to disk. 
<p>
This function should be called after loading a dynalign save file or after a dynalign calculation has been performed. This function generates no error flag. Nothing can go wrong... <dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>filename</em>&nbsp;</td><td>is the file to which the alignment should be written. </td></tr>
  </table>
</dl>

</div>
</div><p>
<hr><h2>Member Data Documentation</h2>
<a class="anchor" name="08dd429e7365f45f0af4ac65f91a38b3"></a><!-- doxytag: member="Dynalign_object::align" ref="08dd429e7365f45f0af4ac65f91a38b3" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">short** <a class="el" href="class_dynalign__object.html#08dd429e7365f45f0af4ac65f91a38b3">Dynalign_object::align</a><code> [private]</code>          </td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>

</div>
</div><p>
<a class="anchor" name="a0ef3ac79a7e95a7f68357cd7dd7d7a2"></a><!-- doxytag: member="Dynalign_object::array" ref="a0ef3ac79a7e95a7f68357cd7dd7d7a2" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double*** <a class="el" href="class_dynalign__object.html#a0ef3ac79a7e95a7f68357cd7dd7d7a2">Dynalign_object::array</a><code> [private]</code>          </td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>

</div>
</div><p>
<a class="anchor" name="57fb75d0ae9ae4d7ddb660ef46073304"></a><!-- doxytag: member="Dynalign_object::ct_templated" ref="57fb75d0ae9ae4d7ddb660ef46073304" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool <a class="el" href="class_dynalign__object.html#57fb75d0ae9ae4d7ddb660ef46073304">Dynalign_object::ct_templated</a><code> [private]</code>          </td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>

</div>
</div><p>
<a class="anchor" name="a1e04d4bab3329f52d69487d6deb87d0"></a><!-- doxytag: member="Dynalign_object::data" ref="a1e04d4bab3329f52d69487d6deb87d0" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">datatable* <a class="el" href="class_dynalign__object.html#a1e04d4bab3329f52d69487d6deb87d0">Dynalign_object::data</a><code> [private]</code>          </td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>

</div>
</div><p>
<a class="anchor" name="51fda66b732a81b71221c1025535a389"></a><!-- doxytag: member="Dynalign_object::dsv_templated" ref="51fda66b732a81b71221c1025535a389" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool <a class="el" href="class_dynalign__object.html#51fda66b732a81b71221c1025535a389">Dynalign_object::dsv_templated</a><code> [private]</code>          </td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>

</div>
</div><p>
<a class="anchor" name="a71db4fdbda2e78a7aad59461ab1e499"></a><!-- doxytag: member="Dynalign_object::forcealign" ref="a71db4fdbda2e78a7aad59461ab1e499" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">short** <a class="el" href="class_dynalign__object.html#a71db4fdbda2e78a7aad59461ab1e499">Dynalign_object::forcealign</a><code> [private]</code>          </td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>

</div>
</div><p>
<a class="anchor" name="4d6d04c0f6e6574db9eb38626aa856e4"></a><!-- doxytag: member="Dynalign_object::gap" ref="4d6d04c0f6e6574db9eb38626aa856e4" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">short <a class="el" href="class_dynalign__object.html#4d6d04c0f6e6574db9eb38626aa856e4">Dynalign_object::gap</a><code> [private]</code>          </td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>

</div>
</div><p>
<a class="anchor" name="85bf2b95d7c490f01c0b3c63d1391696"></a><!-- doxytag: member="Dynalign_object::highend" ref="85bf2b95d7c490f01c0b3c63d1391696" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">short * <a class="el" href="class_dynalign__object.html#85bf2b95d7c490f01c0b3c63d1391696">Dynalign_object::highend</a><code> [private]</code>          </td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>

</div>
</div><p>
<a class="anchor" name="aeb0bf4251ffad85d88791409b74438e"></a><!-- doxytag: member="Dynalign_object::lowend" ref="aeb0bf4251ffad85d88791409b74438e" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">short* <a class="el" href="class_dynalign__object.html#aeb0bf4251ffad85d88791409b74438e">Dynalign_object::lowend</a><code> [private]</code>          </td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>

</div>
</div><p>
<a class="anchor" name="46b200ee4d625bf5b93dd25ee80282e5"></a><!-- doxytag: member="Dynalign_object::lowest" ref="46b200ee4d625bf5b93dd25ee80282e5" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">short <a class="el" href="class_dynalign__object.html#46b200ee4d625bf5b93dd25ee80282e5">Dynalign_object::lowest</a><code> [private]</code>          </td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>

</div>
</div><p>
<a class="anchor" name="fdf4fc676f616371f56e4c9fef9df01f"></a><!-- doxytag: member="Dynalign_object::MAXDSV" ref="fdf4fc676f616371f56e4c9fef9df01f" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">float <a class="el" href="class_dynalign__object.html#fdf4fc676f616371f56e4c9fef9df01f">Dynalign_object::MAXDSV</a><code> [private]</code>          </td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>

</div>
</div><p>
<a class="anchor" name="26f77a6eaa11a89728b53e96c09d25e5"></a><!-- doxytag: member="Dynalign_object::Maxtrace" ref="26f77a6eaa11a89728b53e96c09d25e5" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int <a class="el" href="class_dynalign__object.html#26f77a6eaa11a89728b53e96c09d25e5">Dynalign_object::Maxtrace</a><code> [private]</code>          </td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>

</div>
</div><p>
<a class="anchor" name="d0016552f28504bf8fb4e514d493e68b"></a><!-- doxytag: member="Dynalign_object::modificationflag" ref="d0016552f28504bf8fb4e514d493e68b" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int <a class="el" href="class_dynalign__object.html#d0016552f28504bf8fb4e514d493e68b">Dynalign_object::modificationflag</a><code> [private]</code>          </td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>

</div>
</div><p>
<a class="anchor" name="092a0c489e8c0cd1453e6c593ac25240"></a><!-- doxytag: member="Dynalign_object::savefileread" ref="092a0c489e8c0cd1453e6c593ac25240" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool <a class="el" href="class_dynalign__object.html#092a0c489e8c0cd1453e6c593ac25240">Dynalign_object::savefileread</a><code> [private]</code>          </td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>

</div>
</div><p>
<a class="anchor" name="ac369239b115d06378f886687438cf46"></a><!-- doxytag: member="Dynalign_object::templatefilename" ref="ac369239b115d06378f886687438cf46" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">char* <a class="el" href="class_dynalign__object.html#ac369239b115d06378f886687438cf46">Dynalign_object::templatefilename</a><code> [private]</code>          </td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>

</div>
</div><p>
<a class="anchor" name="8ca9c12e9cb5f53dac734e150d5b39be"></a><!-- doxytag: member="Dynalign_object::v" ref="8ca9c12e9cb5f53dac734e150d5b39be" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">varray* <a class="el" href="class_dynalign__object.html#8ca9c12e9cb5f53dac734e150d5b39be">Dynalign_object::v</a><code> [private]</code>          </td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>

</div>
</div><p>
<a class="anchor" name="6fcebe057ab0c01d452f0f9101494da5"></a><!-- doxytag: member="Dynalign_object::vmod" ref="6fcebe057ab0c01d452f0f9101494da5" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">dynalignarray * <a class="el" href="class_dynalign__object.html#6fcebe057ab0c01d452f0f9101494da5">Dynalign_object::vmod</a><code> [private]</code>          </td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>

</div>
</div><p>
<a class="anchor" name="f4e94d7ad0c260badd19bc00c9c9c9b6"></a><!-- doxytag: member="Dynalign_object::w" ref="f4e94d7ad0c260badd19bc00c9c9c9b6" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">dynalignarray* <a class="el" href="class_dynalign__object.html#f4e94d7ad0c260badd19bc00c9c9c9b6">Dynalign_object::w</a><code> [private]</code>          </td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>

</div>
</div><p>
<a class="anchor" name="a2809c6e4cb309f6598a9f3cc93268f8"></a><!-- doxytag: member="Dynalign_object::w3" ref="a2809c6e4cb309f6598a9f3cc93268f8" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">wendarray * <a class="el" href="class_dynalign__object.html#a2809c6e4cb309f6598a9f3cc93268f8">Dynalign_object::w3</a><code> [private]</code>          </td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>

</div>
</div><p>
<a class="anchor" name="9703f79b7dd3d5349ffcad85a02a87b5"></a><!-- doxytag: member="Dynalign_object::w5" ref="9703f79b7dd3d5349ffcad85a02a87b5" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">wendarray* <a class="el" href="class_dynalign__object.html#9703f79b7dd3d5349ffcad85a02a87b5">Dynalign_object::w5</a><code> [private]</code>          </td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>

</div>
</div><p>
<hr>The documentation for this class was generated from the following files:<ul>
<li>C:/Users/dhm/Documents/RNAstructure57/RNAstructure/RNA_class/<a class="el" href="_dynalign__object_8h.html">Dynalign_object.h</a><li>C:/Users/dhm/Documents/RNAstructure57/RNAstructure/RNA_class/<a class="el" href="_dynalign__object_8cpp.html">Dynalign_object.cpp</a></ul>
</div>
<hr size="1"><address style="text-align: right;"><small>Generated on Fri Nov 28 02:13:17 2014 for RNAstructure Classes by&nbsp;
<a href="http://www.doxygen.org/index.html">
<img src="doxygen.png" alt="doxygen" align="middle" border="0"></a> 1.5.7.1 </small></address>
</body>
</html>
